narwhal/primary/src/certificate_waiter.rs

```diff
@@ -3,226 +3,381 @@
+            current_round < target_round
+        });
+        if self.target.is_empty() {
+            trace!("Certificates have caught up. Skip fetching.");
+            return;
+        }
 
-            // Either upon time-out or round change
-            if attempt_garbage_collection {
-                let round = *self.rx_consensus_round_updates.borrow();
-                if round > self.gc_depth {
-                    let gc_round = round - self.gc_depth;
-
-                    self.pending.retain(|_digest, (r, once_cancel)| {
-                        if *r <= gc_round {
-                            // note: this send can fail, harmlessly, if the certificate has been delivered (`notify_read`)
-                            // and the present code path fires before the corresponding `waiting` item is unpacked above.
-                            let _ = once_cancel
-                                .take()
-                                .expect("This should be protected by a write lock")
-                                .send(());
-                            false
-                        } else {
-                            true
-                        }
-                    });
+        let handle = self.create_fetch_task(progression);
+        let epoch = self.committee.epoch();
+        let metrics = self.metrics.clone();
+        let tx_kick_from_fetch = self.tx_kick_from_fetch.clone();
+
+        self.fetch_certificates_task = Some(tokio::task::spawn(async move {
+            metrics
+                .certificate_waiter_inflight_fetch
+                .with_label_values(&[&epoch.to_string()])
+                .set(1);
+            metrics
+                .certificate_waiter_fetch_attempts
+                .with_label_values(&[&epoch.to_string()])
+                .inc();
+
+            let now = Instant::now();
+            match handle.await {
+                Ok(_) => {
+                    debug!("Finished task to fetch certificates successfully");
                 }
-            }
+                Err(e) => {
+                    debug!("Finished task to fetch certificates with error: {e}");
+                }
+            };
 
-            self.update_metrics(waiting.len());
+            metrics
+                .certificate_waiter_op_latency
+                .with_label_values(&[&epoch.to_string()])
+                .observe(now.elapsed().as_secs_f64());
+            metrics
+                .certificate_waiter_inflight_fetch
+                .with_label_values(&[&epoch.to_string()])
+                .set(0);
+
+            // Schedule another fetch task to check progression and fetch again if needed.
+            let _ = tx_kick_from_fetch.send(()).await;
+        }));
+
+        debug!("Started task to fetch certificates");
+    }
+
+    #[allow(clippy::mutable_key_type)]
+    fn create_fetch_task(
+        &mut self,
+        progression: BTreeMap<PublicKey, Round>,
+    ) -> JoinHandle<DagResult<()>> {
```

Comment:
why make this a second spawned task with a JoinHandle, instead of an async fn that is awaited in fetch_certificates_task?

Reply:
Was trying to manage the captured variables. Refactored to have only one spawned task.